userapi/storage/postgres/threepid_table.go

```diff
@@ -33,21 +34,22 @@ CREATE TABLE IF NOT EXISTS userapi_threepids (
 	medium TEXT NOT NULL DEFAULT 'email',
 	-- The localpart of the Matrix user ID associated to this 3PID
 	localpart TEXT NOT NULL,
+	server_name TEXT NOT NULL,
 
 	PRIMARY KEY(threepid, medium)
 );
 
-CREATE INDEX IF NOT EXISTS userapi_threepid_idx ON userapi_threepids(localpart);
+CREATE INDEX IF NOT EXISTS userapi_threepid_idx ON userapi_threepids(localpart, server_name);
 `
 
 const selectLocalpartForThreePIDSQL = "" +
-	"SELECT localpart FROM userapi_threepids WHERE threepid = $1 AND medium = $2"
+	"SELECT localpart, server_name FROM userapi_threepids WHERE threepid = $1 AND medium = $2"
```

Comment:
I think we need to update this to also filter on the `server_name` column, as otherwise this might return more than one result.